### PR TITLE
Metametrics

### PR DESCRIPTION
**Pull request by danjm**
This PR implements the analytics system described here https://medium.com/metamask/metametrics-an-introduction-to-metamasks-new-analytics-system-e9749e74a275

Code related to the sending of metrics can be reviewed. Reviewers may find it easy to review commit by commit.

Outstanding UI tasks include:

- [x] Add correct chart graphic on opt-in screen
- [x] Update the opt-in modal (for existing users) to latest designs
- [x] Apply the responsive design to the opt-in modal
- [x] Make other first time flow changes on which this will depend
- [x] Rebase this onto first time flow changes
- [x] Update tests

peek 2019-02-28 13-04

### CODE REVIEW

**Comment by danjm** [comment]
Now with responsive designs

peek 2019-02-18 23-15

**Comment by danjm** [comment]
Will rebase onto https://github.com/MetaMask/metamask-extension/pull/6192 once merged

File: test/e2e/beta/from-import-beta-ui.spec.js [hunk 1]
```diff
@@ -110,6 +112,12 @@ describe('Using MetaMask with an existing account', function () {
...
 
+    it('clicks the "No thanks" option on the metametrics opt-in screen', async () => {
+      const optOutButton = await findElement(driver, By.css('.btn-default'))
+      optOutButton.click()
+      await delay(largeDelayMs)
+    })
+
     it('imports a seed phrase', async () => {
       const [seedTextArea] = await findElements(driver, By.css('textarea.first-time-flow__textarea'))
       await seedTextArea.sendKeys(testSeedPhrase)
```
Comment (tmashuang) [inline test/e2e/beta/from-import-beta-ui.spec.js]: screen shot 2019-02-28 at 10 50 17 am

The TOS needs to be clicked after the seed phrase and password is entered.

**Comment by tmashuang** [comment]
https://github.com/MetaMask/metamask-extension/blob/cb2698d20eae273d372d03e11fa765a91c330c17/test/e2e/beta/from-import-beta-ui.spec.js#L123


Might need to be 
`const tosCheckBox = await findElement(driver, By.css('.first-time-flow__checkbox-label'))
`

**Comment by danfinlay** [comment]
Noticed two things:

1. When unlocking and then clicking out of the popup, the metrics prompt was not shown again until my next log in (maybe is fine).

2. I was unable to scroll down to the buttons to approve this. There were two scroll wheels visible, neither let me get to the buttons:

metametrics-bug1

**Comment by danfinlay** [comment]
Oh, and is this page going to link to any of our more detailed blog posts on the issue? I think that would be nice.

**Comment by bdresser** [comment]
Looks good @danjm! Couple notes on these changes
* "privacy policy here" on opt-in step should link [here](https://metamask.io/privacy.html)
* Opt-in screen is a little text-heavy Can we remove the last line of text in the lead paragraph  ("If you care about the Usability....")
* Can we make the legalese underneath the YES/NO buttons a medium shade of gray
* I can't check the ToS checkbox on the Import Seed step, getting the below error
screen shot 2019-02-28 at 10 54 17 am
* I can't change the gas default to "Fast" or "Slow" and getting this error 
screen shot 2019-02-28 at 11 18 14 am

Willl keep playing around and check coorresponding behavior in Matomo!

Also, I didn't properly QA the changes to first-time flow from https://github.com/MetaMask/metamask-extension/pull/6192. That's my bad. Adding other things I noticed in https://github.com/MetaMask/metamask-extension/issues/6222

**Comment by danjm** [comment]
@bdresser I fixed the errors with import ToS and the fast and slow buttons on the send screen.

**Comment by bdresser** [comment]
When I organically start the send flow, then clilck "Next" to get to the confirm screen, the extension is popping open a standalone confirm modal in addition to the popup (still attached to Chrome). So there are two MM's showing the same confirmation screen

**Comment by bdresser** [comment]
"Clicked Create Account" is coming through to Matomo as "Clicked Create Accoount" which is 100% my bad, I typo'd in the spec. Updated in spec.

File: ui/app/actions.js [hunk 2]
```diff
@@ -2607,6 +2615,47 @@ function toggleAccountMenu () {
   }
 }
 
+function setParticipateInMetaMetrics (val) {
+  return (dispatch) => {
+    log.debug(`background.setParticipateInMetaMetrics`)
+    return new Promise((resolve, reject) => {
+      background.setParticipateInMetaMetrics(val, (err, metaMetricsId) => {
+        console.log('!@# err, metaMetricsId', err, metaMetricsId)
```
Comment (whymarrh) [inline ui/app/actions.js]: We can leave this in if we switch to using `log.debug`

File: ui/app/actions.js [hunk 3]
```diff
@@ -2607,6 +2615,47 @@ function toggleAccountMenu () {
...
+        if (err) {
+          dispatch(actions.displayWarning(err.message))
+          return reject(err)
+        }
+
+        resolve([val, metaMetricsId])
+      })
+      dispatch({
```
Comment (whymarrh) [inline ui/app/actions.js]: Should this action be dispatched inside the callback for `MetaMaskController#setParticipateInMetaMetrics`?

File: ui/app/actions.js [hunk 4]
```diff
@@ -2607,6 +2615,47 @@ function toggleAccountMenu () {
...
+    return new Promise((resolve, reject) => {
+      background.setMetaMetricsSendCount(val, (err) => {
+        if (err) {
+          dispatch(actions.displayWarning(err.message))
+          return reject(err)
+        }
+
+        resolve(val)
+      })
+      dispatch({
```
Comment (whymarrh) [inline ui/app/actions.js]: Similar note to above

File: ui/app/actions.js [hunk 5]
```diff
@@ -2693,3 +2742,18 @@ function clearApprovedOrigins () {
...
+
+function setFirstTimeFlowType (type) {
+  return (dispatch) => {
+    log.debug(`background.setFirstTimeFlowType`)
+    background.setFirstTimeFlowType(type, (err) => {
+      if (err) {
+        return dispatch(actions.displayWarning(err.message))
+      }
+    })
+    dispatch({
```
Comment (whymarrh) [inline ui/app/actions.js]: Similar note to above

File: ui/app/app.js [hunk 6]
```diff
@@ -82,6 +83,22 @@ class App extends Component {
     if (!currentCurrency) {
       setCurrentCurrencyToUSD()
     }
+
+    this.props.history.listen((locationObj, action) => {
+      if (action === 'PUSH') {
+        const url = `&url=${encodeURIComponent('http://www.metamask.io/metametrics' + locationObj.pathname)}`
+        this.context.metricsEvent({}, {
+          // ...props,
```
Comment (whymarrh) [inline ui/app/app.js]: Nit: can we remove this line

File: ui/app/app.js [hunk 7]
```diff
@@ -82,6 +83,22 @@ class App extends Component {
...
+          // ...config,
```
Comment (whymarrh) [inline ui/app/app.js]: Nit: can we remove this line

File: ui/app/components/button/button.component.js [hunk 8]
```diff
@@ -18,7 +18,7 @@ const typeHash = {
   raised: CLASSNAME_RAISED,
   'first-time': CLASSNAME_FIRST_TIME,
 }
-
+//
```
Comment (whymarrh) [inline ui/app/components/button/button.component.js]: Nit: can we remove this line?

File: ui/app/components/pages/confirm-transaction-base/confirm-transaction-base.component.js [hunk 9]
```diff
@@ -348,6 +400,29 @@ export default class ConfirmTransactionBase extends Component {
           })
       }
     })
+
+    setMetaMetricsSendCount(metaMetricsSendCount + 1)
+
+    if (onSubmit) {
+      Promise.resolve(onSubmit(txData))
```
Comment (whymarrh) [inline ui/app/components/pages/confirm-transaction-base/confirm-transaction-base.component.js]: I don't think we need to wrap this in a `Promise` chain, as these all seem like sync calls. Could we write:

```js
try {
    onSubmit(...)
    this.setState(...)
} catch (e) {
    // ....
}
```

File: ui/app/components/sidebars/sidebar.component.js [hunk 10]
```diff
@@ -14,10 +14,16 @@ export default class Sidebar extends Component {
...
     type: PropTypes.string,
     sidebarProps: PropTypes.object,
+    onOverlayClose: PropTypes.func,
   };
 
   renderOverlay () {
-    return <div className="sidebar-overlay" onClick={() => this.props.hideSidebar()} />
+    const { onOverlayClose } = this.props
+
+    return <div className="sidebar-overlay" onClick={() => {
```
Comment (whymarrh) [inline ui/app/components/sidebars/sidebar.component.js]: Nit: can we put the props for this `div` on new lines? (And format them etc.)

File: ui/app/metametrics/metametrics.provider.js [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+  getChildContext () {
+    const props = this.props
+    const { pathname } = location
+    const { previousPath, currentPath } = this.state
+
+    return {
+      metricsEvent: (config, overrides = {}) => {
+        const isSendFlow = Boolean(config.eventOpts && config.eventOpts.name && config.eventOpts.name.match(/^send|^confirm/) || overrides.pathname && overrides.pathname.match(/send|confirm/))
+
+        // if (userPermission) {
```
Comment (whymarrh) [inline ui/app/metametrics/metametrics.provider.js]: Nit: can we remove this line?

File: ui/app/metametrics/metametrics.provider.js [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+  }
+
+  getChildContext () {
+    const props = this.props
+    const { pathname } = location
+    const { previousPath, currentPath } = this.state
+
+    return {
+      metricsEvent: (config, overrides = {}) => {
+        const isSendFlow = Boolean(config.eventOpts && config.eventOpts.name && config.eventOpts.name.match(/^send|^confirm/) || overrides.pathname && overrides.pathname.match(/send|confirm/))
```
Comment (whymarrh) [inline ui/app/metametrics/metametrics.provider.js]: Nit: can we introduce some extra variable (or something) to shorten this line a bit?

File: ui/app/metametrics/metametrics.provider.js [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+      },
+    }
+  }
+
+  render () {
+    return this.props.children
+  }
+}
+
+MetaMetricsProvider.propTypes = {
```
Comment (whymarrh) [inline ui/app/metametrics/metametrics.provider.js]: We should put this as a static property on the class directly

File: ui/app/metametrics/metametrics.provider.js [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+  network: PropTypes.string,
+  environmentType: PropTypes.string,
+  activeCurrency: PropTypes.string,
+  accountType: PropTypes.string,
+  metaMetricsSendCount: PropTypes.number,
+  children: PropTypes.object,
+  history: PropTypes.object,
+}
+
+MetaMetricsProvider.childContextTypes = {
```
Comment (whymarrh) [inline ui/app/metametrics/metametrics.provider.js]: We should put this as a static property on the class directly

File: ui/app/metametrics/metametrics.provider.js (hunk 13 above)
Comment (whymarrh) [reply ui/app/metametrics/metametrics.provider.js]: Also, (totally just bringing this up now, but) if we know that any of these props are necessary for this component to do its thing, we can tack on `.isRequired` to the prop type.

File: ui/app/metametrics/metametrics.util.js [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+  [METAMETRICS_CUSTOM_NUMBER_OF_TOKENS]: 9,
+  [METAMETRICS_CUSTOM_NUMBER_OF_ACCOUNTS]: 10,
+}
+
+function composeUrlRefParamAddition (previousPath, confirmTransactionOrigin) {
+  const externalOrigin = confirmTransactionOrigin && confirmTransactionOrigin !== 'MetaMask'
+  return `&urlref=${externalOrigin ? 'EXTERNAL' : encodeURIComponent(previousPath.replace(/chrome-extension:\/\/\w+/, 'http://www.metamask.io/metametrics'))}`
+}
+
+// function composeActionNameParamAddition (pathname, pageOpts) {
```
Comment (whymarrh) [inline ui/app/metametrics/metametrics.util.js]: Nit: can we remove this comment?

File: ui/app/metametrics/metametrics.util.js [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+  const base = METAMETRICS_BASE_FULL
+
+  const e_c = composeParamAddition(eventOpts.category, 'e_c')
+  const e_a = composeParamAddition(eventOpts.action, 'e_a')
+  const e_n = composeParamAddition(eventOpts.name, 'e_n')
+  const new_visit = isNewVisit ? `&new_visit=1` : ''
+
+  const cvar = customVariables && composeCustomVarParamAddition(customVariables) || ''
+
+  const action_name = '' // pageOpts && pathname && composeActionNameParamAddition(pathname, pageOpts) || ''
```
Comment (whymarrh) [inline ui/app/metametrics/metametrics.util.js]: Nit: do we need this comment?

File: ui/app/selectors.js [hunk 17]
```diff
@@ -50,6 +56,46 @@ function getNetworkIdentifier (state) {
...
 }
 
+function getCurrentKeyring (state) {
+  const identity = getSelectedIdentity(state)
+
+  if (!identity) {
+    return null
+  }
+
+  const simpleAddress = identity.address.substring(2).toLowerCase()
```
Comment (whymarrh) [inline ui/app/selectors.js]: Is there a helper function anywhere to strip a hex prefix?

**Review by whymarrh — CHANGES_REQUESTED** [review]
A few things, mostly nitpicks

**Comment by danjm** [comment]
- [ ] Don't send a second opt-in event if the user goes back
- [ ] Collect events on going back and on selecting create vs import after going back
- [ ] ensure all custom variables are supported (i.e. that they have mappings to custom variable ids)

**Comment by bdresser** [comment]
I think opting IN to metrics is being recorded as opting out

**Comment by danjm** [comment]
> I think opting IN to metrics is being recorded as opting out

Good catch again. Fixed now.

**Review by whymarrh — APPROVED** [review]

**Pull request merged**